lib/src/layer/tile_layer/tile_painter.dart

```diff
@@ -0,0 +1,62 @@
+import 'package:flutter/material.dart';
+import 'package:flutter/rendering.dart';
+import 'package:flutter_map/src/layer/tile_layer/tile_model.dart';
+
+/// Draws [TileModel]s onto a canvas at the correct position
+class TilePainter extends CustomPainter {
+  /// List of [TileModel]s to draw to the canvas
+  List<TileModel> tiles;
+
+  /// Paint to use when drawing each tile
+  ///
+  /// Defaults to [defaultTilePaint].
+  final Paint? tilePaint;
```

Comment:
Does `tilePaint` give enough flexibility? `TileBuilder` used to give loads of flexibility, and users could do basically anything they wanted. We might want to consider what users want to do a bit more.

Reply:
I am not sure what would be missing comparing with TileBuilder. User could get image from tile and modify it?

Reply:
Modifying an incoming tile's bytes can be done with a `TileProvider`, so there's no gap there - but previously you could 'overlay'/'wrap' widgets, etc. See https://demo.fleaflet.dev/tile_builder.

I understand that positioning widgets over this kind of defeats the point of a canvas, and is one of the reasons we previously didn't use canvas, to allow this kind of easy tweaking. However, we should really provide the ability to overlay widgets at tile positions. Note that it's not possible to paint a widget to a canvas.

Reply:
I see. I looked into it, then there is now way to avoid Stack. I suppose could then preserve Stack and Tile approach and use that if `TileBuilder` is not null else use CustomPainter?

Reply:
Not necessarily. I suppose a good compromise would be to allow a point in the painting process that can be influenced by the user, ie. they can paint to the same canvas, but are constrained to a smaller `Size`, for each tile. Although much less convienient for the user, this will allow as much flexibility as possible.

I guess it's also worth verifying those performance improvements on a few different devices, just to see if we're making a correct decision here, and the pros outweigh the cons, IYSWIM.

Reply:
tile_builder has been nice to debug tile loading. I wouldn't prefer to have widgets mixed in with the canvas. Most if not everything that has been done with widgets should be possible directly on the canvas too. 
In addition to the `tilePaint` parameter I'd suggest to add a callback that let's users draw custom things on the canvas that gets called after the tiles draws.